### PR TITLE
approval-distribution: Update topology if authorities are discovered later

### PR DESCRIPTION
**Pull request by alexggh**
Fixes: https://github.com/paritytech/polkadot-sdk/issues/2138.

Especially on restart AuthorithyDiscovery cache is not populated so we create an invalid topology  and messages won't be routed correctly for the entire session. This PR proposes to try to fix this by updating the topology as soon as we now the Authority/PeerId mapping, that should impact the situation dramatically.

[This issue was hit yesterday](https://grafana.teleport.parity.io/goto/o9q2625Sg?orgId=1), on Westend and resulted in stalling the finality.

# TODO

- [x] Unit tests
- [x] Test impact on versi

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4951978

File: polkadot/node/network/protocol/src/grid_topology.rs [hunk 1]
```diff
@@ -89,6 +89,25 @@ impl SessionGridTopology {
...
+		ids: &HashSet<AuthorityDiscoveryId>,
+	) -> bool {
+		if self.peer_ids.contains(&peer_id) {
+			for peer in self
+				.canonical_shuffling
+				.iter_mut()
+				.filter(|peer| ids.contains(&peer.discovery_id))
+			{
+				peer.peer_ids.push(peer_id);
+				self.peer_ids.insert(peer_id);
```
Comment (eskimor) [inline polkadot/node/network/protocol/src/grid_topology.rs:105]: If it is already in peer_ids, why do we insert it?

Comment (alexggh) [reply polkadot/node/network/protocol/src/grid_topology.rs:105]: Yeah, that's why `tests` it is on the todo list.

I was looking more about what you think of the approach does it seem reasonable to you ?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4952400

File: polkadot/node/network/protocol/src/grid_topology.rs (hunk 1 above)
Comment (eskimor) [reply polkadot/node/network/protocol/src/grid_topology.rs:105]: Seems reasonable at first sight, yes.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4989543

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4990507

**Comment by alexggh** [comment]
Tested on versi and this PR has the expected behaviour, at node restart we don't have any authority-peer_id mapping in the case, but at next active leaves updates all Peer ids are correctly detected.

**Review by ordian — APPROVED** [review]
Looks reasonable. Nice work!

File: polkadot/node/network/protocol/src/grid_topology.rs [hunk 2]
```diff
@@ -291,6 +312,25 @@ impl SessionGridTopologyEntry {
...
+		peer_id: PeerId,
+		ids: &HashSet<AuthorityDiscoveryId>,
+	) -> bool {
+		let peer_id_updated = self.topology.update_authority_ids(peer_id, ids);
+		// If we added a new peer id we need to recompute the grid neighbors, so that
+		// neighbors_x and neighbors_y reflect the right peer ids.
+		if peer_id_updated {
+			if let Some(local_index) = self.local_index.as_ref() {
+				let grid_neighbors =
+					self.topology.compute_grid_neighbors_for(*local_index).unwrap();
```
[the remark this answers is not in the document]
Comment (alexggh) [reply polkadot/node/network/protocol/src/grid_topology.rs]: Good catch we don't need the unwraps.

**Review by sandreim — APPROVED** [review]

**Pull request merged**